runtime/src/main/java/org/corfudb/runtime/view/stream/BackpointerStreamView.java

```diff
@@ -0,0 +1,367 @@
+        if (getCurrentContext().logPointer > maxAddress) {
+            return;
+        }
+
+        // First, we fetch the current token from the sequencer.
+        final long latestToken = runtime.getSequencerView()
+                .nextToken(Collections.singleton(streamID), 0)
+                .getToken();
+
+        // The highest queued read is either the log pointer if there are no
+        // elements on the list or the maximum element in the read queue.
+        final long highestQueuedRead =
+                getCurrentContext().readQueue.size() == 0 ?
+                getCurrentContext().logPointer :
+                getCurrentContext().readQueue.last();
+
+        // Now we start traversing backpointers, if they are available. We
+        // start at the latest token and go backward, until we reach the
+        // previous highestQueuedRead. For each address which is less than
+        // maxAddress, we insert it into the read queue.
+        long currentRead = latestToken;
+
+        while (currentRead > highestQueuedRead) {
+            // Read the entry in question.
+            LogData currentEntry =
+                    runtime.getAddressSpaceView().read(currentRead);
+
+            // If the current entry is unwritten, we need to fill it,
+            // otherwise we cannot resolve the stream.
+            if (currentEntry.getType() == DataType.EMPTY) {
+
+                // We'll retry the read a few times, we should only need
+                // to fill if a client has actually failed, which should
+                // be a relatively rare event.
+                for (int i = 0; i < NUM_RETRIES; i++) {
+                    currentEntry =
+                            runtime.getAddressSpaceView().read(currentRead);
+                    if (currentEntry.getType() != DataType.EMPTY) {
+                        break;
+                    }
+                }
+
+                // If we STILL don't have the data, now we need to do a hole
+                // fill.
+                if (currentEntry.getType() == DataType.EMPTY) {
+                    try {
+                        runtime.getAddressSpaceView().fillHole(currentRead);
+                        // If we reached here, our hole fill was successful.
+                    } catch (OverwriteException oe) {
+                        // If we reached here, this means the remote client
+                        // must have successfully completed the write and
+                        // we can continue.
+                    }
+                }
+
+                // At this point the hole is filled or has the data and we
+                // can continue.
+            }
+
+            // If the entry contains this context's stream,
+            // and it is less than max read, we add it to the read queue.
+            if (currentEntry.containsStream(contextStreamID) &&
+                    currentRead <= maxAddress) {
+                getCurrentContext().readQueue.add(currentRead);
```

Comment:
So , technically, the read-queue is really a stack .. the last position added is the first read.